### ISSUE TITLE
Export digest of built container image

### ISSUE THREAD

**Issue by mxey** [earthly/earthly#1258]
I would like Earthly to programatically provide the digest of a container image it built, so I could reference it in other steps/targets to exactly identify an image for deployment.

I don't know how this would look, maybe another EARTHLY_ build arg?

**Comment by mologie** [comment]
Was about to file the same issue :)

There are two use cases here:

1. Storing the digest of an image that was built via `--image`
2. Accessing the digest of an image that was built through another step

Kaniko is a single-step build tool and covers use-case 1 nicely: https://github.com/GoogleContainerTools/kaniko#--image-name-with-digest-file

Earthly could mirror Kaniko's flags, e.g. `earthly --image-name-with-digest-file foo.digest --image +foo` to have a deployment step read from `foo.digest` and inject this exact digest into a K8s deployment file.

The second use-case, which seems to be your @mxey primary driver, does not seem to fit all too well into the command collection we have in Earthly right now.

We'd somehow have to create a dependency between the stage that wants the digest and the source stage/image, but neither `FROM` nor `COPY`, which create those dependencies right now, seem suitable for the job.

Best thing I could come up with on short notice is a new `DIGEST` keyword, using the behavior of `ARG` but with a target reference as parameter.

I'd suggest an option `--env` to have it use the behavior of `ENV` instead of `ARG` and persist the the value as env variable into the layer. Additionally, we could mirror Kaniko's options to control the value format (image+digest, or image+tag+digest).


```
FROM ubuntu:20.04

foo:
    RUN touch /bar
    SAVE IMAGE foo:latest

deploy-foo:
    DIGEST EXAMPLE1=+foo
    DIGEST --env EXAMPLE2=+foo
    DIGEST --env --with-tag EXAMPLE2=+foo
    RUN echo $EXAMPLE1 # prints foo@sha256:cbbf2...
    RUN echo $EXAMPLE2 # like EXAMPLE1
    RUN echo $EXAMPLE3 # prints foo:latest@sha256:cbbf2...
    SAVE IMAGE foo/deploy:latest

bar:
    FROM deploy-foo
    RUN echo $EXAMPLE1 # prints nothing
    RUN echo $EXAMPLE2 # prints foo@sha256:cbbf2...
```

**Comment by mxey** [comment]
RUN cd deploy && kustomize edit set image foo:latest $EXAMPLE1
    SAVE IMAGE foo/deploy:latest

I think this will not work because the digest cannot be known before the image is created. The digest covers image metadata like its timestamp.

Maybe something like this:

```
image:
    ...

deploy:
   BUILD --set-image-var=IMAGE +image 
   RUN kustomize edit set image foo:latest $IMAGE

**Comment by mxey** [comment]
> I believe you misunderstood my example. The SAVE IMAGE part you quoted is irrelevant for the RUN line above it, it merely demonstrates propagation of digests into stages descending from deploy-foo, specifically bar.

You are correct. I was confused because of the `SAVE IMAGE` in the deploy target.

**Comment by vladaionescu** [comment]
Possible workaround?

```
the-target:
  ...
  SAVE IMAGE image-name:latest

id:
  WITH DOCKER --load +the-target
    RUN docker inspect --format='{{.RepoDigests}}' image-name:latest >id.txt
  END
  SAVE ARTIFACT id.txt
```

Note that this requires tweaking as it comes out wrapped in `[...]`.

**Comment by mologie** [comment]
I considered that but docker's `images` command is really bad for that use case. Most of the time there is no digest until you push the image via dockerd itself, and with `--format` there is a long-open bug where targeting a single image breaks it completely.

https://github.com/docker/cli/issues/728

```
$ docker image ls --format '{{.Digest}}'
sha256:e1c082e3d3c45cccac829840a25941e679c25d438cc8412c2fa221cf1a824e6a
$ docker image ls --format '{{.Digest}}' alpine:latest
<none>
```

Since Earthly does its own pushing this workaround (combined with shell hackery to get around the digest filter bug) only works for some use cases, but not for Earthly's `--ci`.

**Comment by kodeninja** [comment]
We have the same use case as well. We leverage `skaffold` for deploying our docker images via GitLab CI, and we need to pass the build artifacts (which were earlier generated via `skaffold build --file-output=tags.json ...`, now replaced by `earthly ...`) to the deploy stage.

Currently, I'm using the following (inefficient) hack to get by:

```bash
$ earthly --image -P +docker
$ docker pull "$CI_REGISTRY_IMAGE:$CI_COMMIT_TAG"
$ IMAGE_TAG=$(docker inspect --format='{{index .RepoDigests 0}}' "$CI_REGISTRY_IMAGE:$CI_COMMIT_TAG" | awk -F'@sha256:' '{print $2}')
$ /bin/cat > tags.json <<EOF
{
  "builds": [
    {
      "imageName": "$CI_REGISTRY_IMAGE",
      "tag": "$CI_REGISTRY_IMAGE:$CI_COMMIT_TAG@sha256:${IMAGE_TAG}"
    }
  ]
}
EOF
```

**Comment by theodiem** [comment]
I'm moving from drone ci to earthly and found this issue. I'm used to pass the artifact file to kaniko as previously mentioned.
Just my 5 cents suggestion, to keep things simple, in my existing pipeline I produce the json from kaniko, in another step I process that json to what future steps would use (lets say a .env file to be sourced) and in another step I produce a manifest from a template which is fed through this env.

I would suggest to keep it "earthly" (pun intended) since there won't be a format that would satisfy all. So as long as there's an artifact file with the information from the image build (name, digest, tags, etc). You can prepare your "expected" input easily with jq/yq/go template in an intermediate step.

But yeah, +1 for the feature to have a "digest" output in a structured format to be saved as artifact

**Comment by MattDelac** [comment]
👋 Any progress on that issue?
When using github actions, it would be nice to have something as handy as the docker GHA that spills out the digest
https://github.com/docker/build-push-action#outputs
